src/program.cpp

```diff
@@ -457,65 +203,8 @@ std::vector<argument> generic_eval(const program& p,
                                    std::unordered_map<std::string, argument> params,
                                    F trace)
 {
-    assert(p.validate() == p.end());
-    std::unordered_map<instruction_ref, argument> results;
-    results.reserve(p.size() * 2);
-    std::vector<argument> values;
-    values.reserve(16);
-    for(auto ins : iterator_for(p))
-    {
-        const auto& name = ins->name();
-        if(name == "@literal")
-        {
-            results.emplace(ins, trace(ins, [&] { return ins->get_literal().get_argument(); }));
-        }
-        else if(name == "@param")
-        {
-            results.emplace(
-                ins, trace(ins, [&] {
-                    auto param_name = any_cast<builtin::param>(ins->get_operator()).parameter;
-                    if(not contains(params, param_name))
-                        MIGRAPHX_THROW("Parameter not found: " + param_name);
-                    auto param = params[param_name];
-                    if(param.get_shape() != ins->get_shape())
-                        MIGRAPHX_THROW("Incorrect shape {" + to_string(param.get_shape()) +
-                                       "} for parameter: " + param_name);
-                    return param;
-                }));
-        }
-        else if(name == "@outline")
-        {
-            results.emplace(ins, trace(ins, [&] { return argument{ins->get_shape(), nullptr}; }));
-        }
-        else if(name == "@return")
-        {
-            std::vector<argument> prog_outputs;
-            std::transform(ins->inputs().begin(),
-                           ins->inputs().end(),
-                           std::back_inserter(prog_outputs),
-                           [&](instruction_ref i) {
-                               assert(results.find(i) != results.end());
-                               return results[i];
-                           });
-
-            return prog_outputs;
-        }
-        else
-        {
-            values.resize(ins->inputs().size());
-            std::transform(
-                ins->inputs().begin(), ins->inputs().end(), values.begin(), [&](instruction_ref i) {
-                    assert(results.find(i) != results.end());
-                    return results[i];
-                });
-            results.emplace(ins, trace(ins, [&] {
-                                return ins->get_operator().compute(ctx, ins->get_shape(), values);
-                            }));
-        }
-        assert(results.find(ins) != results.end());
-    }
-
-    return {results.at(std::prev(p.end()))};
+    const auto* mm = p.get_main_module();
+    return generic_eval(*mm, ctx, params, trace);
```

Comment:
There shouldn't be a `generic_eval` in program. Instead `eval` from module should be called:

```cpp
std::vector<argument> program::eval(parameter_map params) const
{
    const auto* mm = p.get_main_module();
    mm->eval(params);
}
```

All the tracing handling should be done in the module.

Reply:
My previous change is like the above and I can change it back

Reply:
Should we keep the generic_eval() in the program.cpp? It is called in a couple of places.

Reply:
Ah sorry, yea just leave it as it is and move generic_eval into program.cpp.